### PR TITLE
Allow params on menu resource items - POC

### LINKED ISSUE #1651: Allow resource menu items to receive params
### Feature

From [this discussion](https://discord.com/channels/740892036978442260/740893011994738751/1090017637079207946).

We'd love it if the `resource` menu items would receive the `params` option, which would be a `Hash`, and that would be appended to the generated URL.

```ruby
resource :clients, label: "Active", params: {
  filters: "THE_FILTER_QUERY"
}
```

As a block
```ruby

        resource :projects, params: -> {
          params = {}

          if current_user.is_admin?
            params[:foo] = :bar
          end

          params
        }
```

### Current workarounds

One can use a `link_to` operator.

```ruby
link_to "Active", "/avo/resource/clients?filters=THE_FILTER_QUERY"
```

### PR DESCRIPTION
**Pull request by bryszard**
# Description

Fixes https://github.com/avo-hq/avo/issues/1651

It was proposed that the `resource` menu builder method will accept `params` keyword which will be then passed to the path that is rendered on the generated URL.

To achieve that I've added one optional option on the `Avo::Menu::Resource` class and used it to pass to the `parsed_resource`, which is then used in `Avo::Sidebar::ItemSwitcherComponent` and passed to `helpers.resources_path(resource: resource)`.

Finally, I've modified `resources_path` helper to merge the `resource_params` into the query params.

There is an side effect of this change - we cannot use anymore the `:inclusive` strategy on the `active_link_to` helper that is highlighting the specific link on sidebar, because then all the items for the same path will be `active`. 

What I've done instead is using the `:exact` strategy. This solves the problem, but also brings another one - now if you add any additional query param to the path, the link on the sidebar won't be highlighted. This might or might not be a problem from the UX perspective. I need your advice in this matter.

How it looks now:

https://github.com/bryszard/avo/assets/12682792/42a88a28-6b30-4e8b-b391-0c949a341ac4

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the [documentation](https://github.com/avo-hq/avodocs)
- [ ] I have added tests that prove my fix is effective or that my feature works

## Manual review steps

1. Run the dummy app with new config included
1. Click through the menu items
1. Apply different filters on the `Users` and `Admins` page 

Manual reviewer: please leave a comment with output from the test if that's the case.

### CODE REVIEW

**Comment by Paul-Bob** [comment]
Thank you for this POC @bryszard we appreciate it and we're truly grateful for your efforts!! After analyzing your approach and our source code I have some suggestions on a slightly different approach that may simplify this feature.

We should start by collecting the `params` that you already did by adding the `option :params, optional: true` to the `lib/avo/menu/resource.rb`.

Now the second step should be to pass that `params` while generating the URL and here is where my suggestion and your approach diverges. If we look on `app/components/avo/sidebar/item_switcher_component.html.erb` there is this piece of code: 

```ruby
<% if item.is_a? Avo::Menu::Resource %>
  <%= render Avo::Sidebar::LinkComponent.new label: item.navigation_label, path: helpers.resources_path(resource: resource), data: item.data, icon: item.icon %>
<% end %>
```

Notice that the path is generated by `helpers.resources_path(resource: resource)` and if we check the `resources_path` method's code we notice that it accept any keyword argument and will append it to the path.

With this in mind, augmenting the path with an additional parameter is straightforward: `helpers.resources_path(resource: resource, filters: "some_filters_string")`. This inclusion automatically appends `filters=some_filters_string` to the query parameters.

From here my suggestion is to create a fetcher method inside  `Avo::Menu::Resource` and use it to append the params while calling the `resources_path` method, something like `helpers.resources_path(resource: resource, item.fetch_params)`

Notice that on the request @adrianthedev mentioned about having the possibility to make the params callable too, for that we use an internal class `Avo::ExecutionContext`, there is some examples of the usage on the code, let me know if need some guidance on how to use it.

I'm eager to hear your thoughts on this proposal!

**Comment by bryszard** [comment]
@Paul-Bob Thanks for the feedback, I'll apply your suggestions.

However, there is still a doubt about the UX - how should we deal with the `active_link_to` and highlighting of the item on the sidebar? What are the solutions I see:
1. we just highlight on the exact match and stop highlighting if any additional param is added (might be good enough, but I'm not sure if it's what you want),
1. we highlight all items that match the route (default, strange and I'd say unwanted UX),
1. we find another custom solution (eg. use some "active" state kept in the local storage in the browser), which might require removing the `active_link_to` library.

**Comment by Paul-Bob** [comment]
There is a strong possibility that `active_link_to` will just work with this new approach.

**Comment by Paul-Bob** [comment]
I just noticed you already committed the suggestions, I think it end up clean and straight to the point, great work @bryszard ! Just tested it out and `active_link_to` seems to work too! Amazing!

File: lib/avo/menu/resource.rb [hunk 1]
```diff
@@ -11,4 +12,8 @@ def parsed_resource
   def entity_label
     parsed_resource.navigation_label
   end
+
+  def fetch_params
+    params || {}
```
Comment (Paul-Bob) [inline lib/avo/menu/resource.rb]: Here we should consider that `params` can be a `proc`. We use [Avo::ExecutionContext](https://docs.avohq.io/3.0/execution-context.html) in this cases. Check how we use it [here](https://github.com/avo-hq/avo/blob/c7ca9b02b0734e1e23f22f95e152e78887b169e1/lib/avo/fields/base_field.rb#L189).

The `target` would be the `params` in this case and the other arguments are what we want to give access to the user inside the executed block.

Comment (bryszard) [reply lib/avo/menu/resource.rb]: Thank you, will do :).

**Comment by bryszard** [comment]
@Paul-Bob I'm not sure if I understand correctly your comment about the active link. This is how it looks now:

image

So you're saying that highlighting multiple items at once is fine, right?

**Comment by Paul-Bob** [comment]
> So you're saying that highlighting multiple items at once is fine, right?

Nope, my bad, didn't noticed it while testing 🤦

We should find a way to fix it

File: lib/avo/menu/resource.rb (hunk 1 above)
Comment (bryszard) [reply lib/avo/menu/resource.rb]: Done.

**Comment by bryszard** [comment]
@Paul-Bob So, one way to deal with the multiple items highlighting is using the `active: :exact` option. Then the behavior looks like this:

https://github.com/bryszard/avo/assets/12682792/0b884962-bb4e-410e-8e48-ac3a2e9234c8

It's not perfect, I know, because when you add any additional param to the path, different then the original one, no item will be highlighted. That includes even the `page` param.

What if...

We generate different paths for those new pages? We could base the path eg. by the `label` or we could add additional parameter to the `resource` method ([here](https://github.com/avo-hq/avo/blob/main/lib/avo/menu/builder.rb#L32)) that would tell how the path should be called.

File: lib/avo/menu/resource.rb (hunk 1 above)
Comment (Paul-Bob) [reply lib/avo/menu/resource.rb]: Amazing, is the `resource` the instantiated on or should we pass `resource: parsed_resource` instead?

**Comment by Paul-Bob** [comment]
> We generate different paths for those new pages? We could base the path eg. by the label or we could add additional parameter to the resource method ([here](https://github.com/avo-hq/avo/blob/main/lib/avo/menu/builder.rb#L32)) that would tell how the path should be called.

Not sure if adding the `param` to the `resource` would solve the problem because it's always the same `resource`.

We have a tricky issue here.

Not sure if `exact` provides the best UX, it will work only if the user doesn't interact with filters / pagination etc...

Let's say we have the same resource declared in 3 different ways:

```ruby
resource "UserResource"
resource "UserResource", label: "Admins", params: { filters: "eyJJc0FkbWluIjpbImFkbWlucyJdfQ" }
resource "UserResource", label: "Active Admins", params: -> do
  { active: true, filters: "eyJJc0FkbWluIjpbImFkbWlucyJdfQ" }
end
```

Can we apply the following logic?

- If `filters: "eyJJc0FkbWluIjpbImFkbWlucyJdfQ" ` param is present than label `Admins` is highlighted, even there is more params....
-  If ` active: true` and `filters: "eyJJc0FkbWluIjpbImFkbWlucyJdfQ"` are both present there `Active Admins` is highlighted. 
- If none of the above params are delcared than `User` is highlighted. 

Make sense?

**Comment by bryszard** [comment]
@Paul-Bob I don't know how to achieve such logic with `active_link_to`. These are the possibilities according [to the documentation](https://github.com/comfy/active_link_to#active-options):

image

The 2 points of your logic should be possible with passing the `params` to `active_link_to`. However, I don't see a way to declare the negative condition and achieve the point 3 of the logic. Each menu item is resolved separately, so there is no place to put an "if... else..." condition.

Coming back to:

> Not sure if adding the param to the resource would solve the problem because it's always the same resource.

So I understand that you want to stay REST-compliant and `User` resource must be under `/users` path? I thought we could customize it and allow it to be set eg. to `/admins` or `/non_admins`.

**Comment by bryszard** [comment]
@Paul-Bob I've decided to show to you what I mean. 

Basically, we would need to allow to create some alternative routes for resources. 
Then we could use those routes on the menu item definitions.

Take a look [on the last commit](https://github.com/bryszard/avo/pull/1/commits/9aaa2cb16aef74ec5b1cbb181ed0a68373607f39).

This is how it behaves:

https://github.com/bryszard/avo/assets/12682792/1650dd4a-750a-4179-88ff-d6bc2207ad73

**Comment by Paul-Bob** [comment]
Sorry for the delay, the alternative_route_keys seems to be a working solution. However I'm not sure about this approach, it works on this particular case but I'm afraid that It can become difficult to maintain. 

I think this particular use case of different resources with a default filter each is already implemented by scopes on Avo3, check [demo users' page](https://main.avodemo.com/avo/resources/users)

image

Maybe we're trying to fix something that is not supposed to be used like this, I think the params should be used for passing extra params to a resource but not necessarly to create several resources with different filters.

Also if an customer really wants this 3 different resources instead applying the scopes feature, that can be achieved by creating each resource that points to `self.model_class = User` and apply a different `resolve_query_scope` on each one

Makes sense?

**Comment by bryszard** [comment]
Yes, I think you're right. It's just not supposed to be used multiple times on the same resource.

Which means that we don't need to fix this problem, great! :)

Sorry for getting us into this rabbit hole. I'll finish the PR and open it for review based on our discussion. Thanks!

**Comment by Paul-Bob** [comment]
> Which means that we don't need to fix this problem, great! :)

Exactly =))

 > Sorry for getting us into this rabbit hole. I'll finish the PR and open it for review based on our discussion. Thanks!

No worries! Thank you, the PR looks great!

**Pull request closed without merging**